### ISSUE TITLE
DatabaseTestCase::testUnion() and DatabaseTestCase::testUnionAll() assume order returned by database.

### ISSUE THREAD

**Issue by jlfranklin** [backdrop/backdrop-issues#5797]
## Description of the bug

When adding data to a (clean?) MySQL database, the data is returned in the order it is inserted absent an ORDER BY clause in the query. For UNION SELECT queries, SQL doesn't support ORDER BY in the individual SELECT queries, only at the end covering the entire UNION.  The database layer doesn't have a way to add an ORDER BY to the UNION, only the individual SELECT queries.

Two tests, DatabaseTestCase::testUnion() and DatabaseTestCase::testUnionAll() sometimes fail on non-MySQL databases because the data comes back in a different order than expected.  Calling `sort()` on the results is sufficient to ensure the tests pass and does not invalidate the tests.

## Steps To Reproduce

Run tests on the platform as shown below
OR
Visual inspection of the two test functions.

## Actual behavior

```
---- Database: Select tests (DatabaseSelectTestCase) ----


Status    Group      Filename          Line    Function                            
------------------------------------------------------------------------------------------------------------------------
Fail      Other      database_test.tes 1562    testUnion()                                                             
    First query returned correct name.
Fail      Other      database_test.tes 1563    testUnion()                                                             
    Second query returned correct name.
7 tests skipped.
```

## Expected behavior

Tests pass.

## Additional information

- Test done with Silkscreen 1.21 & PostgreSQL
- PHP 7.4 
- Nginx
- Debian 11.

**Comment by indigoxela** [comment]
@jlfranklin many thanks for reporting and for the PR. Tricky ... you might have guessed - I can't test it with B.

If I get it right, the order is *by accident* correct in MySQL/MariaDB. Wouldn't that mean that the assert in these two tests is misleading, anyway? Shouldn't that rather be an "in_array()" and/or a count of "array_values()"?
(Sorry, I'm not too familiar with UNION altogether.)

**Comment by indigoxela** [comment]
> The call to assertEqual() passes a count of an array.

@argiepiano unless I'm missing something, the assertEqual() checks for a specific value by a specific array index:

```
$this->assertEqual($names[0], 'George', 'First query returned correct name.');
```
But in an unsorted $names array "George" might also be in `$names[1]`, which makes that check unreliable with PostgreSQL - it seems to work as expected in MySQL/MariaDB, but only because these return the values sorted as they got inserted.
So to my understanding the assert text "First/Second query returned correct name." is a bit misleading.
(Note: I'm really no expert - especially not with PostgreSQL.)

**Comment by argiepiano** [comment]
You are right in the case you outline here @indigoxela, but the code actually looks like this:

```php
$this->assertEqual(count($names), 3, 'UNION ALL correctly preserved duplicates.');
```

So, this statement is actually passing the result of count (a number), and comparing it to another number (3). The order of the array items is irrelevant to `count()`:

`count(array('cat', 'dog')) == 2`
and 
`count(array('dog', 'cat')) == 2`

The order of the array items doesn't change the count

**Comment by argiepiano** [comment]
Never mind. I was missing the rest of the code. Was only looking at the code shown in the PR.

**Comment by jlfranklin** [comment]
> But in an unsorted $names array "George" might also be in $names[1], which makes that check unreliable with PostgreSQL - it seems to work as expected in MySQL/MariaDB, but only because these return the values sorted as they got inserted.

@indigoxela, that is correct.  I've seen this happen in other tests and we've added ORDER BY clauses to the SELECT queries to make the test more robust.  (We don't want to rely on a quirk of MySQL's for a test to pass.)  Here, that's not an option, so the PR sorts the returned array instead.

**Comment by indigoxela** [comment]
How can we proceed here? The PR looks reasonable enough to merge, but it fixes a problem we can't actually reproduce in B.
As it affects our affiliated project Silkscreen, I'm leaning towards merging. Do people agree?

**Comment by indigoxela** [comment]
FTR, we have a meta with [Drupal crossports](https://github.com/backdrop/backdrop-issues/issues/5770), and there's a UNION-related issue: https://www.drupal.org/node/2772107

@jlfranklin would that patch have any impact on these tests?

**Comment by jlfranklin** [comment]
> @jlfranklin would that patch have any impact on these tests?

It would let us use an ORDER BY to sort the results instead of `sort()`.  Other than that, no, I don't see any material changes from that patch.
